infra.bs

```diff
@@ -872,6 +886,156 @@ For <a>pairs</a>, a slightly shorter literal syntax can be used, separating the
 as 200/`<code>OK</code>`.
 
 
+<h2 id=forgiving-base64>Forgiving base64</h2>
+
+<p>To <dfn export>forgiving-base64 encode</dfn> given a <a>byte sequence</a> <var>data</var>, apply
+the base64 algorithm defined in section 4 of RFC 4648 to <var>data</var> and return the result.
+[[!RFC4648]]
+
+<p class="note no-backref">This is named <a>forgiving-base64 encode</a> for symmetry with
+<a>forgiving-base64 decode</a>, which is different from the RFC as it defines error handling for
+certain inputs.
+
+<p>To <dfn export>forgiving-base64 decode</dfn> given a string <var>data</var>, run these steps:</p>
+
+<ol>
+ <li><p>Let <var>position</var> be a <a>position variable</a> for <var>data</var>, initially
+ pointing at the start of <var>data</var>.
+
+ <li><p>Remove all <a>ASCII whitespace</a> from <var>data</var>.
+
+ <li>
+  <p>If <var>data</var> contains a code point that is not one of
+
+  <ul class="brief">
+   <li>U+002B (+)
+   <li>U+002F (/)
+   <li><span>ASCII alphanumeric</span>
+  </ul>
+
+  <p>then return failure.
+
+ <li>
+  <p>If <var>data</var>'s <a for=string>length</a> divides by 4 leaving no remainder, then:
+
+  <ol>
+   <li><p>If <var>data</var> ends with one or two U+003D (=) characters, then remove them from
+   <var>data</var>.
+  </ol>
+
+ <li><p>If <var>data</var>'s <a for=string>length</a> divides by 4 leaving a remainder of 1, then
+ return failure.
+
+ <li><p>Let <var>output</var> be an empty <a>byte sequence</a>.
+
+ <li><p>Let <var>buffer</var> be an empty buffer that can have bits appended to it.
+
+ <li>
+  <p>While <var>position</var> does not point past the end of <var>data</var>:
+
+  <ol>
+   <li>
+    <p>Find the character pointed to by <var>position</var> in the first column of the following
+    table. Let <var>n</var> be the number given in the second cell of the same row.</p>
+
+    <div id="base64-table">
```

Comment:
We could alternately say second column ... first cell of the same row in the RFC 4648 table 1.